### PR TITLE
Added methods to blockmask to visualize them

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129915
* __->__ #129950
* #129846

image
image

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/nn/attention/_flex_attention.py [hunk 1]
```diff
@@ -139,7 +196,7 @@ def _create_block_mask_from_mask(
         q_indices=q_indices.to(torch.int32).to(mask.device).contiguous(),
         KV_BLOCK_SIZE=KV_BLOCK_SIZE,
         Q_BLOCK_SIZE=Q_BLOCK_SIZE,
-    ).as_tuple()
+    )
```
Comment (yanboliang) [inline torch/nn/attention/_flex_attention.py:200]: You want to update the func's return value annotation at L124.

**Review by yanboliang — APPROVED** [review]

File: torch/_inductor/kernel/flex_attention.py [hunk 2]
```diff
@@ -387,10 +387,10 @@ def build_subgraph_buffer(
     (torch.float32, 128): (128, 32, 4, 3),
     (torch.float32, 256): (64, 16, 4, 3),
     (torch.bfloat16, 64): (128, 64, 4, 3),
-    (torch.bfloat16, 128): (128, 128, 8, 2),
+    (torch.bfloat16, 128): (128, 64, 8, 3),
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:390]: was this just more tuning?

File: torch/nn/attention/_flex_attention.py [hunk 3]
```diff
@@ -76,6 +76,63 @@ def as_tuple(self):
...
+            valid_indices = torch.where(index_mask, kv_indices, num_cols)
+
+            # set the values in 'a' to 1 where the indices are valid
+            dense_mask[row_indices, valid_indices] = 1
+            return dense_mask[:, :num_cols]
+
+        out = create_dense_one(self.kv_num_blocks[0, 0], self.kv_indices[0, 0])
+        return out
+
+    def to_string(self, max_rows=20, max_cols=20):
```
Comment (drisspg) [inline torch/nn/attention/_flex_attention.py:112]: maybe add some descriptions on when and how you would use these?

**Review by drisspg — APPROVED** [review]

File: torch/_inductor/kernel/flex_attention.py (hunk 2 above)
Comment (Chillee) [reply torch/_inductor/kernel/flex_attention.py:390]: Yeah. Also, to avoid the segfaults I'm avoiding `num_stages=2`.

**Pull request closed without merging**